common/rpc/encryption/localStoreCertProvider.go

```diff
@@ -78,73 +113,102 @@ func (s *localStoreCertProvider) FetchServerCertificate() (*tls.Certificate, err
 }
 
 func (s *localStoreCertProvider) FetchClientCAs() (*x509.CertPool, error) {
-	if s.tlsSettings.Server.ClientCAFiles == nil {
+	if len(s.tlsSettings.Server.ClientCAFiles) == 0 && len(s.tlsSettings.Server.ClientCAData) == 0 {
 		return nil, nil
 	}
 
-	// Check under a read lock first
 	s.RLock()
 	if s.clientCAs != nil {
 		defer s.RUnlock()
 		return s.clientCAs, nil
 	}
-	// Not found, manually unlock read lock and move to write lock
+
 	s.RUnlock()
 	s.Lock()
 	defer s.Unlock()
 
-	var clientCaPool *x509.CertPool
-	if len(s.tlsSettings.Server.ClientCAFiles) > 0 {
-		var err error
-		clientCaPool, err = buildCAPool(s.tlsSettings.Server.ClientCAFiles)
-		if err != nil {
-			return nil, err
-		}
+	if s.clientCAs != nil {
+		return s.clientCAs, nil
+	}
+
+	clientCaPool, err := buildCAPool(s.tlsSettings.Server.ClientCAFiles, s.tlsSettings.Server.ClientCAData)
+	if err != nil {
+		return nil, err
 	}
 
 	s.clientCAs = clientCaPool
 	return s.clientCAs, nil
 }
 
 func (s *localStoreCertProvider) FetchServerRootCAsForClient() (*x509.CertPool, error) {
-	if s.tlsSettings.Client.RootCAFiles == nil {
+	if len(s.tlsSettings.Client.RootCAFiles) == 0 && len(s.tlsSettings.Client.RootCAData) == 0 {
 		return nil, nil
 	}
-	// Check under a read lock first
+
 	s.RLock()
 	if s.serverCAs != nil {
 		defer s.RUnlock()
-		return s.clientCAs, nil
+		return s.serverCAs, nil
 	}
-	// Not found, manually unlock read lock and move to write lock
+
 	s.RUnlock()
 	s.Lock()
 	defer s.Unlock()
 
-	var serverCAPool *x509.CertPool
-	if len(s.tlsSettings.Client.RootCAFiles) > 0 {
-		var err error
-		serverCAPool, err = buildCAPool(s.tlsSettings.Client.RootCAFiles)
-		if err != nil {
-			return nil, err
-		}
+	if s.serverCAs != nil {
+		return s.serverCAs, nil
+	}
+
+	serverCAPool, err := buildCAPool(s.tlsSettings.Client.RootCAFiles, s.tlsSettings.Client.RootCAData)
+	if err != nil {
+		return nil, err
 	}
 
 	s.serverCAs = serverCAPool
 	return s.serverCAs, nil
 }
 
-func buildCAPool(caFiles []string) (*x509.CertPool, error) {
+func buildCAPool(caFiles []string, caData []string) (*x509.CertPool, error) {
+	atLeastOneCert := false
 	caPool := x509.NewCertPool()
+
 	for _, ca := range caFiles {
+		if ca == "" {
```

Comment:
did this so that we are not forced to create pass redundant variables as we did in the tls sample